Add tracks events to a few sections

octaedro opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:
This PR adds the code to record a few tracks events that we didn't have in the new product creation experience.

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #35456.

<!-- The next section is mandatory. If your PR doesn't require testing, please indicate that you are purposefully omitting instructions. -->

- [ ] This PR is a very minor change/addition and does not require testing instructions (if checked you can ignore/remove the next section).

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Otherwise, please include detailed instructions on how these changes can be tested (including pre-conditions, configuration, steps to take and expected results). It may help to write your instructions using pseudocode -- as if you're telling a computer how to execute the test. -->

Request:
1.  Install the [WC-Admin Test Helper](https://github.com/woocommerce/woocommerce-admin-test-helper) and go to `Tools > WCA Test Helper > Tools` and `Enable wc-admin Tracking`. Also enable the `new-product-management-experience`.
2. Go to `Products > Add New (MVP)`.
3. Go to the Images section and press `Choose Images`. Select an image and confirm that the event `wcadmin_product_image_add_from_media_library` is being recorded.
4. Select the image that has been added and press `Replace`. Replace the image with another. The event `wcadmin_product_image_replace` should be recorded.
5. Select the image and press the trash can and check that the event `wcadmin_product_image_change_order` is recorded.
6. Drag and drop an image and confirm that the event `wcadmin_product_image_upload` has been recorded.
7. Add (or choose from the media library) more than 2 images and change the images' order and confirm that the event `wcadmin_product_image_change_order` is recorded.
8. Select an image (not the cover image) and select it as the cover image. Verify that the event `wcadmin_product_image_select_as_cover` is being recorded. You'll see that the event `wcadmin_product_image_change_order` will be recorded as well because the order of the images changed.
9. Go to the `Shipping` section and select `Add new shipping class`. A modal will be visible, press `Add` and confirm that the event `wcadmin_product_modal_new_shipping_class_add_button` is recorded.
10. Verify that the cancel button is being recorded too (event name: `wcadmin_product_modal_new_shipping_class_cancel_button`)
10. Now press the link `global settings` and verify that the event `wcadmin_product_shipping_global_settings` is being recorded.
11. Go to `Attributes` and press `Add first attribute`. The event `wcadmin_product_add_first_attribute_button` should be recorded.
12. Verify that the buttons Add (`wcadmin_product_modal_attribute_add_button`) and Cancel (`wcadmin_product_modal_attribute_cancel_button`) are being recorded correctly.
13. Verify that the event `wcadmin_product_add_another_attribute_button` is being recorded after pressing `+ Add another`
14. Select an `Attribute` and then press the trash can. Verify the event `wcadmin_product_remove_attribute_button` is recorded.

<!-- End testing instructions -->

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

octaedro committed:
Add tracks events

octaedro committed:
Add changelog

octaedro committed:
Add more tracks events

# Conflicts:
#	plugins/woocommerce-admin/client/products/fields/attribute-field/add-attribute-modal.tsx
#	plugins/woocommerce-admin/client/products/sections/product-shipping-section.tsx

octaedro committed:
Fix name

octaedro committed:
Fix duplicated event

octaedro committed:
Add select as cover event

octaedro committed:
Add cancel shipping class creation

mattsherman commented on plugins/woocommerce-admin/client/products/sections/product-shipping-section.tsx:
```diff
@@ -347,12 +353,20 @@ export function ProductShippingSection( {
 						} )
 					}
-					onCancel={ () => setShowShippingClassModal( false ) }
+					onCancel={ () => {
```
`onCancel` gets called whenever the `AddNewShippingClassModal` is closed. So, the `product_modal_new_shipping_class_cancel_button` Tracks event fires when the `Add` button is clicked as well.

I think the proper solution would be to open up a new PR to address this, and modify `AddNewShippingClassModal` to accept both `onRequestClose` and `onCancel` props. That way, consumers of that modal can specifically handle when the `Cancal` button is explicitly clicked.

octaedro replied:
Nice catch! I removed the `product_modal_new_shipping_class_cancel_button ` event, and I'll create [a new issue](https://github.com/woocommerce/woocommerce/issues/35574) to handle that in a follow-up PR.

mattsherman commented on plugins/woocommerce-admin/client/products/fields/attribute-field/add-attribute-modal.tsx:
```diff
@@ -81,6 +82,7 @@ export const AddAttributeModal: React.FC< CreateCategoryModalProps > = ( {
 			value: AttributeForm[ keyof AttributeForm ]
 		) => void
 	) => {
+		recordEvent( 'product_remove_attribute_button' );
```
Suggestion: `product_add_attributes_modal_remove_attribute_button_click`

mattsherman commented on plugins/woocommerce-admin/client/products/fields/attribute-field/add-attribute-modal.tsx:
```diff
@@ -270,9 +272,12 @@ export const AddAttributeModal: React.FC< CreateCategoryModalProps > = ( {
-									}
+									onClick={ () => {
+										recordEvent(
+											'product_add_another_attribute_button'
```
Suggestion: `product_add_attributes_modal_add_another_attribute_button_click`

mattsherman commented on plugins/woocommerce-admin/client/products/fields/attribute-field/attribute-field.tsx:
```diff
@@ -29,6 +30,7 @@ export const AttributeField: React.FC< AttributeFieldProps > = ( {
 	const [ showAddAttributeModal, setShowAddAttributeModal ] =
 		useState( false );
 	const onRemove = ( attribute: ProductAttribute ) => {
+		recordEvent( 'product_remove_attribute_button' );
```
Suggestion: `product_remove_attribute_button_click`

octaedro replied:
I removed this event in favor of the comment you made below.

mattsherman commented on plugins/woocommerce-admin/client/products/fields/attribute-field/attribute-field.tsx:
```diff
@@ -29,6 +30,7 @@ export const AttributeField: React.FC< AttributeFieldProps > = ( {
+		recordEvent( 'product_remove_attribute_button' );
 		// eslint-disable-next-line no-alert
 		if ( window.confirm( __( 'Remove this attribute?', 'woocommerce' ) ) ) {
 			onChange( value.filter( ( attr ) => attr.id !== attribute.id ) );
```
Since the user can cancel the removal, we should also track if the confirmed or canceled the removal...

`product_remove_attribute_confirmation_confirm_click` and
`product_remove_attribute_confirmation_cancel_click`

octaedro replied:
Good idea! Fixed in the commit [6dd9945](https://github.com/woocommerce/woocommerce/pull/35262/commits/6dd9945b3b01dc2a7543c8e936e283a2406e9fd0)

mattsherman commented on plugins/woocommerce-admin/client/products/fields/attribute-field/attribute-field.tsx:
```diff
@@ -50,6 +52,7 @@ export const AttributeField: React.FC< AttributeFieldProps > = ( {
 					return newAttr;
 				} ),
 		] );
+		recordEvent( 'product_modal_attribute_add_button' );
```
Suggestion: `product_add_attributes_modal_add_button_click`

mattsherman commented on plugins/woocommerce-admin/client/products/fields/attribute-field/attribute-field.tsx:
```diff
@@ -76,18 +79,24 @@ export const AttributeField: React.FC< AttributeFieldProps > = ( {
-								}
+								onCancel={ () => {
+									recordEvent(
+										'product_modal_attribute_cancel_button'
```
Suggestion: `product_add_attributes_modal_cancel_button_click`

mattsherman commented on plugins/woocommerce-admin/client/products/fields/attribute-field/attribute-field.tsx:
```diff
@@ -76,18 +79,24 @@ export const AttributeField: React.FC< AttributeFieldProps > = ( {
-								}
+								onClick={ () => {
+									recordEvent(
+										'product_add_first_attribute_button'
```
Suggestion: `product_add_first_attribute_button_click`

mattsherman commented on plugins/woocommerce-admin/client/products/fields/attribute-field/attribute-field.tsx:
```diff
@@ -167,14 +176,20 @@ export const AttributeField: React.FC< AttributeFieldProps > = ( {
 				<AddAttributeModal
-					onCancel={ () => setShowAddAttributeModal( false ) }
+					onCancel={ () => {
+						recordEvent( 'product_modal_attribute_cancel_button' );
```
Suggestion: `product_add_attributes_modal_cancel_button_click`

Also, since this same Tracks event name is used above, it should be stored in a `const` to ensure that the same event name is always used.

octaedro replied:
Good idea! I added it in the commit [66d1d13](https://github.com/woocommerce/woocommerce/pull/35262/commits/66d1d131bc7ed26f6fd284b1ed030fd40b4a20fb)

mattsherman commented on plugins/woocommerce-admin/client/products/sections/images-section.tsx:
```diff
@@ -48,6 +48,7 @@ export const ImagesSection: React.FC = () => {
 				( file ) => file.id === parseInt( image?.props?.id, 10 )
 			);
 		} );
+		recordEvent( 'product_image_change_order' );
```
Actually, can we track specifically when the user actually clicks the change image order arrow buttons? If so...

Suggestion: `product_images_change_image_order_button_click`

If not, then `product_images_change_image_order_via_image_gallery`

octaedro replied:
After taking a glance at the package, it seems like it's not possible to track only when the button is clicked. We have an `onOrderChange` event that is triggered after clicking the button or after dropping an image to change the order. 
I renamed the event in the commit [806a256](https://github.com/woocommerce/woocommerce/pull/35262/commits/806a25636222dabbd36210b6dafe71413dc6cfbe)

mattsherman commented on plugins/woocommerce-admin/client/products/sections/images-section.tsx:
```diff
@@ -100,6 +101,7 @@ export const ImagesSection: React.FC = () => {
 						} }
 						onDragEnd={ () => {
 							if ( isRemoving && draggedImageId ) {
+								recordEvent( 'product_image_remove' );
```
Suggestion: `product_images_remove_image_button_click`

mattsherman commented on plugins/woocommerce-admin/client/products/sections/images-section.tsx:
```diff
@@ -119,9 +121,13 @@ export const ImagesSection: React.FC = () => {
 								) === undefined
 							) {
 								images[ replaceIndex ] = media as Image;
+								recordEvent( 'product_image_replace' );
```
Suggestion: `product_images_replace_image_button_click`

mattsherman commented on plugins/woocommerce-admin/client/products/sections/images-section.tsx:
```diff
@@ -119,9 +121,13 @@ export const ImagesSection: React.FC = () => {
 							}
 						} }
+						onSelectAsCover={ () =>
+							recordEvent( 'product_image_select_as_cover' )
```
Suggestion: `product_images_select_image_as_cover_button_click`

mattsherman commented on plugins/woocommerce-admin/client/products/sections/images-section.tsx:
```diff
@@ -169,6 +175,9 @@ export const ImagesSection: React.FC = () => {
 											) === undefined
 										) {
+											recordEvent(
+												'product_image_add_from_media_library'
```
Suggestion: `product_images_add_via_media_library`

mattsherman commented on plugins/woocommerce-admin/client/products/sections/images-section.tsx:
```diff
@@ -177,6 +186,9 @@ export const ImagesSection: React.FC = () => {
 									onUpload={ ( files ) => {
 										if ( files[ 0 ].id ) {
+											recordEvent(
+												'product_image_upload'
```
Suggestion: `product_images_add_via_drag_and_drop_upload`

mattsherman commented on plugins/woocommerce-admin/client/products/sections/product-shipping-section.tsx:
```diff
@@ -208,6 +209,11 @@ export function ProductShippingSection( {
 												type="external"
+												onClick={ () => {
+													recordEvent(
+														'product_shipping_global_settings'
```
Suggestion: `product_shipping_global_settings_link_click`

mattsherman commented on plugins/woocommerce-admin/client/products/sections/product-shipping-section.tsx:
```diff
@@ -347,12 +353,20 @@ export function ProductShippingSection( {
 							Promise< ProductShippingClass >
 						>( shippingClassValues ).then( ( value ) => {
+							recordEvent(
+								'product_modal_new_shipping_class_add_button'
```
Suggestion: `product_new_shipping_class_modal_add_button_click`

mattsherman commented on plugins/woocommerce-admin/client/products/sections/product-shipping-section.tsx:
```diff
@@ -347,12 +353,20 @@ export function ProductShippingSection( {
-					onCancel={ () => setShowShippingClassModal( false ) }
+					onCancel={ () => {
+						recordEvent(
+							'product_modal_new_shipping_class_cancel_button'
```
Suggestion: `product_new_shipping_class_modal_cancel_button_click`

mattsherman reviewed (dismissed):
Thanks for adding all these Tracks events, @octaedro! 🙌

There are a few non-naming related comments.

But, most of my comments are in regards to the event names not adhering to the convention of:

`<source>_<context>_<optional subcontext>_<action>_<optional qualifier>`

from PCYsg-4S2-p2

octaedro committed:
Rename tracks events

# Conflicts:
#	plugins/woocommerce-admin/client/products/sections/product-shipping-section.tsx

octaedro committed:
Add const for tracks event name

octaedro committed:
Rename event

octaedro committed:
Rename events

octaedro committed:
Remove `product_modal_new_shipping_class_cancel_button` trancks event

octaedro committed:
Rename `product_add_first_attribute_button_click` event

octaedro committed:
Fix method

octaedro committed:
Fix confirmation modal events

octaedro committed:
Rename `product_add_attributes_modal_add_button_click` event

octaedro commented:
Hey @mattsherman,
I just addressed the changes that you suggested, could you take another look at this PR?

joshuatf commented on plugins/woocommerce-admin/client/products/fields/attribute-field/add-attribute-modal.tsx:
```diff
@@ -284,9 +287,12 @@ export const AddAttributeModal: React.FC< AddAttributeModalProps > = ( {
-									}
+									onClick={ () => {
+										recordEvent(
+											'product_add_attributes_modal_add_another_attribute_button_click'
```
Just thinking out loud here, I think we need to have a larger conversation around tracks names.  My thinking is that generally we should go from least specific to most specific with the action verb at the end.  This example almost matches that, but might be more like:

`product_attributes_create_modal_add_button_click`

Which almost mirrors the folder structure as well.  We're in the `products/` folder, under the `attributes` field, under the creation modal (updated to be less ambigious with "add"), with the item we're referring to (add button), and the action (click).

mattsherman replied:
I like this suggestion... it is less wordy than the suggestion I made, and therefore easier to read. I think we should update the Tracks event names in this PR to follow this tweak, rather than opening up a follow-up PR (because, ideally we don't change Tracks event names once they are in place).

> My thinking is that generally we should go from least specific to most specific with the action verb at the end

That's the guideline. :-)

joshuatf reviewed (commented):
Thanks for cleaning up all of these @octaedro!

I left a comment in general about how we name events.  It's a small subtle difference, but this is definitely up for debate and I don't want to block the work done in this PR.

I'll leave the final ✅  up to @mattsherman since he originally reviewed this one.

mattsherman commented:
@octaedro @joshuatf I think we should make the naming tweak @joshuatf mentioned.

AnnaMag commented:
@octaedro we can considered this PR ready for merging.

AnnaMag reviewed (approved):
LGTM.
